unix/modffi: Let RV64 pass FFI tests again.

agatti opened the pull request:
### Summary

PR #15800 combined with #15767 would trigger an incompatibility in the libffi version used by the CI environment for testing the RISC-V 64 Unix port.  Using a source-built libffi library would also bring up a RISC-V 64 specific issue when handling unsigned 32 bit values in certain cases.

The first problem is solved by updating the vendored libffi version to v3.6.4 (the latest stable at this point), and pointing the libffi submodule to the official libffi repository.  This, along with some FFI callback memory management changes, closes #15528.

The second problem stems from how RISC-V 64 holds 32-bit integers inside a 64 bits wide register.  Quoting from the "RISC-V Instruction Set Manual Volume 1", section 4.2:

> The compiler and calling convention maintain an invariant that all 32-bit values are held in a sign-extended format in 64-bit registers. Even 32-bit unsigned integers extend bit 31 into bits 63 through 32. Consequently, conversion between unsigned and signed 32-bit integers is a no-op, as is conversion from a signed 32-bit integer to a signed 64-bit integer.

The compiler can handle this since it knows what is going to happen to values whose lifetime it manages, and can generate code that takes this situation into account.  However, for FFI functions the compiler has no idea on what's in store for the value and so it just leaves it alone.  The modffi code wasn't aware of this and would happily return 64-bit signed values if the `uint32_t` value returned by the function had its most significant bit set.

### Testing

Testing occurred in a CI-like environment using the same commands used in `ci.sh` to build both the Qemu Unix RISC-V 64, Arm, and MIPS ports.

agatti committed:
tools/ci.sh: Let RV64 use a source-built version of libffi.

This commit lets the RV64 port use the version of libffi that is bundled
as a submodule in the MicroPython source tree, as the packaged libffi
library coming from Ubuntu's RISC-V repository trashes foreign function
call results on exit.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
unix/modffi: Restrict uint32_t values to 32 bits.

This commit clears the upper 32 bits of returned `uint32_t` values,
which are handled as `unsigned int`s by the MicroPython runtime and
thus could be extended to 64 bits on some platforms.

RV64 holds 32-bit values as signed integers when held in registers, but
the code handling the FFI unsigned int case did not take this into
account.  That introduced test failures when a 32-bit value had its most
significant bit set, as when performing the value extension from 32 to
64 bits, the upper half of the value would be filled with ones.

On 32 bit platforms this change should be converted to a no-op, and
on other 64 bit platforms that aren't RISC-V it shouldn't hurt as the
value being manipulated is expected to only hold valid bits in its lower
half.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

dpgeorge commented:
> The second problem stems from how RISC-V 64 holds 32-bit integers inside a 64 bits wide register. Quoting from the "RISC-V Instruction Set Manual Volume 1", section 4.2:
> 
> > The compiler and calling convention maintain an invariant that all 32-bit values are held in a sign-extended format in 64-bit registers. Even 32-bit unsigned integers extend bit 31 into bits 63 through 32. Consequently, conversion between unsigned and signed 32-bit integers is a no-op, as is conversion from a signed 32-bit integer to a signed 64-bit integer.
> 
> The compiler can handle this since it knows what is going to happen to values whose lifetime it manages, and can generate code that takes this situation into account. However, for FFI functions the compiler has no idea on what's in store for the value and so it just leaves it alone. The modffi code wasn't aware of this and would happily return 64-bit signed values if the `uint32_t` value returned by the function had its most significant bit set.

Wow, very interesting, and great debugging!


> The alternative is to disable FFI tests for RISC-V 64.

I think now that you've done the work and have the FFI tests working, we should keep them enabled.  After all, it was these tests that picked up the above issue with signed/unsigned.  That's exactly why tests and large test coverage is important.

Great work!

dpgeorge commented:
Reviewing this PR in more detail, I think we should first fix #15528 and updated the libffi submodule to v3.4.6.  That will make things much simpler here.  Otherwise we'll just be adding workarounds on workarounds.

agatti committed:
lib/libffi: Update libffi to 3.4.6.

Libffi is updated to the latest stable version (v3.4.6) as the version
of libffi that was linked by the repository predates the introduction of
RISC-V processors (both 32 and 64 bits).

This is necessary as the packaged version of libffi used by the CI for
RISC-V 64 turned out to simply not work correctly, whilst a source-built
version of libffi worked just fine.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
gitignore: Add ffi_lib.so to the gitignore list.

When testing the Unix port with FFI tests enabled, the FFI helper
library that's built as part of the testing process is not removed after
the tests session end.

This commit adds the built library to the gitignore list, so it doesn't
mark the sourcetree as dirty if a test run for the Unix port was started
in the past.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
unix/modffi: Clean up FFI closures memory management.

This commit removes custom FFI closures alloc/free functions, in favour
of using the tracked allocation facility to allocate memory for FFI
callback objects.

This stems from linking issues in the Arm port when updating LibFFI to
the latest stable version, as the overridden alloc/free functions didn't
replace LibFFI's (unlike in other ports).  The original code did no
effective cleanup for allocated callback objects, so there is no real
impact when switching allocation strategy.

The tracked allocation feature used to be enabled only if the Bluetooth
stack integration was enabled.  This commit also enables tracked
allocation support if FFI support is enabled.

Co-authored-by: Damien George <damien@micropython.org>
Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti commented:
Fair point, I've updated libffi and changed the callbacks' memory management code.  This should be now ready for merging, as all tests pass with the new libffi.

dpgeorge commented:
Thanks for updating, this looks really good now, simpler and we have a more recent libffi version.

I tested it with the micropython-lib `signal` module, both with a dynamically linked libffi and in `MICROPY_STANDALONE=1` mode, and the ffi callback works.

dpgeorge commented:
Also good to see that the unix port did not grow in size, it actually shrunk by 6 bytes.  So the original motivation to have custom `ffi_closure_alloc` functions seems to no longer be relevant, at least for the `MICROPY_STANDALONE=0` build (see c9598604c6872d87b6c818c92efc4bce357d988c).
